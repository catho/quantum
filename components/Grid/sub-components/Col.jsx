import PropTypes from 'prop-types';
import styled from 'styled-components';
import { query, calcGutter } from './shared';
import { theme as defaultTheme } from '../../shared';
import { CSSVariables } from '../../GlobalStyle';

const columnGrid = ({ size, offset }) => {
  const offsetStyle = offset ? `${offset + 1}/` : '';

  return `grid-column: ${offsetStyle} span ${size || 12};`;
};

const columnGridLess = ({ size, offset, breakpoint, noGutters, gutter }) => {
  const calculedWidth = size ? (100 / 12) * size : 100;
  const calculedOffset = offset ? (100 / 12) * offset : 100;
  const calculedGutter = calcGutter(
    CSSVariables({
      theme: {
        gutter,
      },
    }).gutter[breakpoint],
    noGutters,
    true,
  );

  const offsetStyle = offset
    ? `
    margin-left: calc(${calculedOffset.toFixed(3)}% + ${
        calculedGutter > 0 ? `(${calculedGutter}px / (12 / ${offset})` : `0px`
      } ) );

    &:first-child {
      margin-left: calc(${calculedOffset.toFixed(3)}% + ${
        calculedGutter > 0 ? `(${calculedGutter}px / (12 / ${offset})` : `0px`
      } ) );
    }
    &:last-child {
      margin-right: 0;
    }
  
  `
    : `
    margin-left: ${
      calculedGutter > 0 ? `calc(${calculedGutter}px / 2)` : '0px'
    };
    margin-right: ${
      calculedGutter > 0 ? `calc(${calculedGutter}px / 2)` : '0px'
    };
    
    &:first-child {
      margin-left: 0;
    }
    &:last-child {
      margin-right: 0;
    }
  `;

  const colWidth =
    calculedGutter > 0
      ? `calc(${calculedWidth.toFixed(
          3,
        )}% - ${calculedGutter}px + (${calculedGutter}px / (12 / ${size ||
          12}) ) )`
      : `${calculedWidth.toFixed(3)}%`;

  return `
    display: inline-block;
    width: ${colWidth};

    ${offsetStyle}
  `;
};

const columnPosition = (
  {
    xsmall,
    small = xsmall,
    medium = small,
    large = medium,
    xlarge = large,
    'xsmall-offset': xsmallOffset,
    'small-offset': smallOffset = xsmallOffset,
    'medium-offset': mediumOffset = smallOffset,
    'large-offset': largeOffset = mediumOffset,
    'xlarge-offset': xlargeOffset = largeOffset,
    theme: { breakpoints, gutter },
    'no-gutters': noGutters,
  },
  breakpoint,
) => {
  const q = query(breakpoints)[breakpoint];

  const screenDefinitions = {
    xsmall: {
      size: xsmall,
      offset: xsmallOffset,
    },
    small: {
      size: small,
      offset: smallOffset,
    },
    medium: {
      size: medium,
      offset: mediumOffset,
    },
    large: {
      size: large,
      offset: largeOffset,
    },
    xlarge: {
      size: xlarge,
      offset: xlargeOffset,
    },
  };

  const { size, offset } = screenDefinitions[breakpoint];

  return q`
    @supports ( display: grid ) {
      ${columnGrid({
        size,
        offset,
        breakpoint,
      })}
    }

    @supports not ( display: grid ) {
      ${columnGridLess({
        size,
        offset,
        breakpoint,
        noGutters,
        gutter,
      })}
    }
  `;
};

const Col = styled.div`
  ${props =>
    Object.keys(props.theme.breakpoints).map(breakpoint =>
      columnPosition(props, breakpoint),
    )}

  word-break: break-word;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
`;

Col.defaultProps = {
  theme: defaultTheme,
  'no-gutters': false,
};

Col.propTypes = {
  'no-gutters': PropTypes.bool,
  xsmall: PropTypes.number,
  small: PropTypes.number,
  medium: PropTypes.number,
  large: PropTypes.number,
  xlarge: PropTypes.number,
  'xsmall-offset': PropTypes.number,
  'small-offset': PropTypes.number,
  'medium-offset': PropTypes.number,
  'large-offset': PropTypes.number,
  'xlarge-offset': PropTypes.number,
  hide: PropTypes.oneOfType([
    PropTypes.oneOf(Object.keys(defaultTheme.breakpoints)),
    PropTypes.arrayOf(PropTypes.oneOf(Object.keys(defaultTheme.breakpoints))),
  ]),
};

Col.displayName = 'Col';

export default Col;
