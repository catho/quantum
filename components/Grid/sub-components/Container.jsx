import PropTypes from 'prop-types';
import styled from 'styled-components';
import { hide, query, calcGutter } from './shared';
import { theme as defaultTheme } from '../../shared';
import { CSSVariables } from '../../GlobalStyle';

const renderBreakpoint = (
  { theme: { gutter, breakpoints }, 'no-gutters': noGutters },
  breakpoint,
) => {
  const calculedGutter = CSSVariables({
    theme: {
      gutter: Number(calcGutter(gutter, noGutters, true)),
    },
  }).gutter[breakpoint];
  const q = query(breakpoints)[breakpoint];

  return q`
    padding: ${calculedGutter};
  `;
};

const renderResponsives = ({
  theme: { breakpoints, gutter },
  'no-gutters': noGutters,
}) =>
  Object.keys(breakpoints).map(breakpoint =>
    renderBreakpoint(
      {
        theme: {
          breakpoints,
          gutter,
        },
        'no-gutters': noGutters,
      },
      breakpoint,
    ),
  );

const Container = styled.div`
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;

  ${renderResponsives}
  ${({ fluid }) => !fluid && 'max-width: 95%;'}
  ${hide}

  & & {
    max-width: 100%;
  }
`;

Container.propTypes = {
  fluid: PropTypes.bool,
  hide: PropTypes.oneOfType([
    PropTypes.oneOf(Object.keys(defaultTheme.breakpoints)),
    PropTypes.arrayOf(PropTypes.string),
  ]),
  theme: PropTypes.shape({
    breakpoints: PropTypes.object,
  }),
  'no-gutters': PropTypes.bool,
};

Container.defaultProps = {
  fluid: false,
  'no-gutters': false,
  theme: defaultTheme,
};

Container.displayName = 'Container';

export default Container;
