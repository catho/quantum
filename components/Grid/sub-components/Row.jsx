import { cloneElement, isValidElement } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Container from './Container';
import styles from './Row.module.css';

const Row = ({
  children,
  'no-gutters': rowNoGutters = false,
  className,
  ...rest
}) => {
  const rowClass = classNames(
    styles.row,
    {
      [styles['no-gutter']]: rowNoGutters,
    },
    className,
  );

  const isNumberOrString = (child) =>
    typeof child === 'string' || typeof child === 'number';

  const applyNoGutters = (child) => {
    if (
      isNumberOrString(child) ||
      (child.type !== Row && child.type !== Container)
    ) {
      return child;
    }

    const { 'no-gutters': childNoGutters } = child.props;

    const ChildWithNoGutters = cloneElement(child, {
      'no-gutters': childNoGutters || rowNoGutters || undefined,
    });
    return ChildWithNoGutters;
  };

  const applyChildrenProps = (c) =>
    Array.isArray(c)
      ? c
          .filter((child) => isValidElement(child) || isNumberOrString(child))
          .map((child) => applyNoGutters(child))
      : applyNoGutters(c);

  return (
    <div {...rest} className={rowClass}>
      {applyChildrenProps(children)}
    </div>
  );
};

Row.propTypes = {
  'no-gutters': PropTypes.bool,
  children: PropTypes.node.isRequired,
};

Row.displayName = 'Row';

export default Row;
