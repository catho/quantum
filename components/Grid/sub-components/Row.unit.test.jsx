import { render, screen } from '@testing-library/react';
import Row from './Row';
import Col from './Col';
import Container from './Container';

const SAMPLE_PHRASE = 'Hello world';

describe('<Row />', () => {
  it('Should render all react nodes', () => {
    const number = 1;

    render(
      <Row>
        <span key={1}>{SAMPLE_PHRASE}</span>
        <Col key={2}>{SAMPLE_PHRASE}</Col>
        {false}
        {/* returned in conditional rendering (ex: condition && <element/>) */}
        {null}
        {/* returned in conditional rendering (ex: condition ? <element/> : null) */}
        {undefined}
        {/* returned when rendering an object's undefined prop (ex: <div{obj.prop}</div>) */}
        {number}
      </Row>,
    );

    const nodeListWithPhraseSample = screen.getAllByText(SAMPLE_PHRASE);
    expect(nodeListWithPhraseSample).toHaveLength(2);

    expect(screen.getByText(/1/)).toBeInTheDocument();
  });

  it('Should forward no-gutters prop only to other Row and Container', () => {
    const { container } = render(
      <Row no-gutters>
        <span key={1}>{SAMPLE_PHRASE}</span>
        <Row key={2}>{SAMPLE_PHRASE}</Row>
        <Container key={3}>{SAMPLE_PHRASE}</Container>
      </Row>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
