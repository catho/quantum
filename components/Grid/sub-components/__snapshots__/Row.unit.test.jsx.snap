// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> Should forward no-gutters prop only to other Row and Container 1`] = `
.Row-module__row___xiasO {
  display: grid;
  grid-auto-columns: max-content;
}

.Row-module__no-gutter___8kj-f {
  margin-bottom: 0!important;
  grid-column-gap: 0!important;
  grid-row-gap: 0!important;
}

.Container-module__container___ZBwIj {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  max-width: 95%;
}

.Container-module__no-gutters___sX06G {
  padding: 0!important;
}

@media (min-width:0px) {
  .Row-module__row___xiasO {
    margin-bottom: var(--qtm-gutter-xsmall);
    grid-column-gap: var(--qtm-gutter-xsmall);
    grid-row-gap: var(--qtm-gutter-xsmall);
    grid-template-columns: repeat(var(--qtm-breakpoint-xsmall-columns),1fr);
  }
}

@media (min-width:600px) {
  .Row-module__row___xiasO {
    margin-bottom: var(--qtm-gutter-small);
    grid-column-gap: var(--qtm-gutter-small);
    grid-row-gap: var(--qtm-gutter-small);
    grid-template-columns: repeat(var(--qtm-breakpoint-small-columns),1fr);
  }
}

@media (min-width:1024px) {
  .Row-module__row___xiasO {
    margin-bottom: var(--qtm-gutter-medium);
    grid-column-gap: var(--qtm-gutter-medium);
    grid-row-gap: var(--qtm-gutter-medium);
    grid-template-columns: repeat(var(--qtm-breakpoint-medium-columns),1fr);
  }
}

@media (min-width:1440px) {
  .Row-module__row___xiasO {
    margin-bottom: var(--qtm-gutter-large);
    grid-column-gap: var(--qtm-gutter-large);
    grid-row-gap: var(--qtm-gutter-large);
    grid-template-columns: repeat(var(--qtm-breakpoint-large-columns),1fr);
  }
}

@media (min-width:0px) {
  .Container-module__container___ZBwIj {
    padding: var(--qtm-gutter-xsmall);
  }
}

@media (min-width:600px) {
  .Container-module__container___ZBwIj {
    padding: var(--qtm-gutter-small);
  }
}

@media (min-width:1024px) {
  .Container-module__container___ZBwIj {
    padding: var(--qtm-gutter-medium);
  }
}

@media (min-width:1440px) {
  .Container-module__container___ZBwIj {
    padding: var(--qtm-gutter-large);
  }
}

<div
  class="Row-module__row___xiasO Row-module__no-gutter___8kj-f"
>
  <span>
    Hello world
  </span>
  <div
    class="Row-module__row___xiasO Row-module__no-gutter___8kj-f"
  >
    Hello world
  </div>
  <div
    class="Container-module__container___ZBwIj Container-module__no-gutters___sX06G"
  >
    Hello world
  </div>
</div>
`;
